README.md
# paperdb
A Conference Proceedings and Papers Web App

Originally created for www.wotug.org in the early 2000s, this is a php/mysql application that implements a conference proceedings and papers
database, and the tools to view and update it. For the wotug.org website it enables access to the papers presented at conferences.

# Installation

See the file Install for more information.

# License

The code is released under the GNU Public License (GPL) version 2. See the file COPYING for more information.


ChangeLog
2014-08-11  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>
  * Moved into GIT

  * Reformatted source using phpstorm

2005-09-27  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* Major update to change parameter used from "id"
	to "num", so that hopefully google will index the
	pages. (it ignores params "id=").

	* Fixed up an sql quoting problem.

	* Made search.php more secure and increased search
	fields it supports.

2004-11-25  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* admin_fns.php: include initialisation of new time
	fields when inserting papers.

	* home_output_fns.php: improve formatting.

2004-11-25  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* Modofied index/follow to be more permissive.

2004-11-18  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* list_proceeds.php: Implemented export to bibtex of
	complete database. Old list_proceeds now f=1, new functions
	are f=2...f=5

	* admin_fns.php, index.php, fixup_text.php: update the
	invocation of list_proceeds.php to include f=1

	* paper_fns.php: implemented routines to get last change
	date of a proceeding or all papers, supporting export.
	
	* show_proc.php: for bibtex export, indicate source of 
	data as being paperdb.

2004-11-17  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* entities.php: ensure that { and } are quoted correctly.

	* refer_output_fns.php: Fixup comma after bibtex key, and 
	don't bracket-quote the key text. Some fixes to the escaping
	code.

2004-11-16  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* <all>: Fixed a number of html parser errors, and
	added html validator check image to footer of my local
	config.php page.

2004-11-15  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>
	
	* <all>: tidied up ampersands in URLs to use &amp;

	* html_output_fns.php: corrected format of DOCTYPE.

	* paper_fns.php: implemented get_comment_by_paper.

	* create_db.sql: additional secondary key field definitions, 
	include accessed, modified date and use created not stored for
	creation date.

	* paper_fns.php: implement group of functions to reset
	access and modified dates and counts.

2004-11-05  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* entities.php: Commented out backslash translation for BibTeX;
	not sure it's really appropriate.

	* refer_output_fns.php: Include new style code (+2) to indicate
	no editor hrefs required.

2004-11-03  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* admin_fns.php: changed length of some of the comment bars. Added
	"Fill in unset refs" command to admin menu, calling the new file
	create_refs.php.

	* admin_fns.php: added "generic_update() to pass on a preformed
	sql update string to the database. Initially used to pass on the
	reftext updates calculated by create_refs.php.

	* entities.php: removed extra newlines after closing php tag.

	* list_authors.php, list_editors.php: add header comment.

	* paper_output_fns.php: Add links to new plain-text biblio ops.

	* refer_output_fns.php: fix swapped-fn bug in do_refer, do_bibtex.
	
	* refer_output_fns.php: Only emit abstract if it's non-null. Fix
	up some <BR>'s in plain-text funcs. 

	* show_proc.php, show_pap.php: integrate code to call plain-text
	functions.
		
2004-10-30  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* refer_output_fns.php: Changed to enable output in text or in
	html form, given help from the translation fns XML2*

	* entities.php: created, including the xml funcs from the
	html_output_fns.php

	* form_output_fns.php, send_file.php, create_db.sql: add new file
	types MS Powerpoint, MS Word.

2004-10-28  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* edit_paper.php: lots of updates to make this work!

2004-10-06  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* admin_fns.php: fix insert_paper to properly treat the proceeding
	parameter as an array!

2004-10-04  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* admin.php: Stopped code re-requesting login details when already
	logged in.

	* admin_fns.php: create delete_paper_file. Change
	insert_paper_file so that the file contents can't be printed with
	an error message.

	* add_file.php: move link_file_to_paper() to admin_fns.php, modify
	some of the messages, add actions menu at end of page

	* show_pap.php, show_proc.php: add new action menu to page.

	* show_pap.php: add in mising require;

	* paper_output_fns.php, form_output_fns.php: include / to end
	single-element tags to start bringing the code towards XHTML
	compatibility.

2004-09-20  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* admin_fns.php: Ensure that all sql strings are properly escaped
	by using new function sqlvalue.

	* db_fns.php: create fn sqlvalue to escape string or numeric value
	properly.

	* edit_proceeding.php: create edit proceeding functions modelled
	on edit_paper.

	* insert_papers.php: check more carefully for errors.

2003-06-23  Ruth Ivimey-Cook  <Ruth.Ivimey-Cook@ivimey.org>

	* add_file.php: remove unnecessary assigns from unset POST vars

	* admin_fns.php: fix up inserts, deletes to authorlist; the field
	name was incorrect

	* edit_paper.php: Major edit so it is now possible to edit most
	details of a paper through the web page, using a 2 stage form.

	* form_output_fns.php: include quoting using htmlspecialchars
	where special characters get in the way.

	* home_output_fns.php: removed extra <p> around abstract

	* list_papers.php: 'Pages' as the title was a bit big: use 'pp'
	instead.

	* paper_fns.php: include 'order by ordering' for accesses to
	authorlist.

	* paper_output_fns.php: display_paper_verbose() now emits an
	'edit' URL in admin mode, and can be asked (via a bool param) to
	display a URL to the proceedings, which is used by show_auth.php.
	New function display_proceedings_details() which displays info
	about a proceedings on one line (para) given proceedings & paper
	records.

	* refer_output_fns.php: include some fn. headers, add 'plain'

search.php
<?php
/**
 * search.php
 *
 * Copyright (c) 2000-2003 Ruth Ivimey-Cook
 * Licensed under the GNU GPL. For full terms see the file COPYING.
 *
 * Display and process the search form.
 *
 * $Id: search.php,v 1.6 2005/09/27 21:33:53 rivimey Exp $
 */

require_once("paper_fns.php");
require_once("html_output_fns.php");
require_once("book_output_fns.php");
require_once("paper_output_fns.php");

session_start();
db_connect(); // for mysql_real_escape_string()

do_html_header("Search results");

if (isset($_POST['goButton']) && $_POST['goButton'] = "Search") {

  $num = $_POST['num'];
  $fr = isset($_POST['fr']) ? $_POST['fr'] : 0;
  $tables = array();
  $wnum = 0;
  $where = array();
  $proc_paper_link = 0;
  $pers_paper_link = 0;
  for ($i = 1; $i <= $num; $i++) {
    if (isset($_POST['what'][$i])) {
      $t = $_POST['text'][$i];
      $w = $_POST['what'][$i];

      if ($w == 1) { // title
        $where[$wnum++] = "papers.title REGEXP " . sqlvalue($t);
        $tables["papers"] = 1;
      }
      elseif ($w == 2) { // author surname
        $where[$wnum++] = "people.lastname REGEXP " . sqlvalue($t);
        $pers_paper_link = 1;
      }
      elseif ($w == 3) { // author firstname
        $where[$wnum++] = "people.firstname REGEXP " . sqlvalue($t);
        $pers_paper_link = 1;
      }
      elseif ($w == 4) { // abstract
        $where[$wnum++] = "papers.abstract REGEXP " . sqlvalue($t, "A");
        $tables["papers"] = 1;
      }
      elseif ($w == 5) { // reftext
        $where[$wnum++] = "papers.reftext REGEXP " . sqlvalue($t);
        $tables["papers"] = 1;
      }
      elseif ($w == 6) { // year
        $where[$wnum++] = "proceedings.pubyear = " . sqlvalue($t, "N");
        $proc_paper_link = 1;
      }
      elseif ($w == 7) { // isbn
        $where[$wnum++] = "proceedings.isbn REGEXP " . sqlvalue($t);
        $proc_paper_link = 1;
      }
      elseif ($w == 8) { // files attached
        $where[$wnum++] = "paperfilelist.paperid = papers.paperid";
        $tables["paperfilelist"] = 1;
        $tables["papers"] = 1;
      }
    }
  }
  if ($proc_paper_link) {
    $where[$wnum++] = "paperlist.proceedingid = proceedings.proceedingid and papers.paperid = paperlist.paperid";
    $tables["paperlist"] = 1;
    $tables["papers"] = 1;
    $tables["proceedings"] = 1;
  }
  if ($pers_paper_link) {
    $where[$wnum++] = "people.personid = authorlist.authorid and authorlist.authors = papers.paperid";
    $tables["papers"] = 1;
    $tables["authorlist"] = 1;
    $tables["people"] = 1;
  }

  // read the tables array and create a comma-separated list of the tables
  // required by the query. Each table is only listed once.
  {
    reset($tables);
    $tablestr = "";
    $done = FALSE;
    while (!$done) {
      $tablename = key($tables);
      $tablestr .= "$tablename";

      $done = (next($tables) == FALSE);
      if (!$done) {
        $tablestr .= ",";
      }
    }
    while (!$done) {
      ;
    }
  }

  // read the where array and turn it into a series of bracketed
  // SQL expressions: '(papers.title REGEXP "string") and ( ... )'
  {
    reset($where);
    $condstr = "";
    do {
      $cond = current($where);
      $condstr .= "( $cond )";

      $done = (next($where) == FALSE);
      if (!$done) {
        $condstr .= " and ";
      }
    } while (!$done);
  }


  // add the preamble to the conditional strings, and run the query. We
  // need to end up with thean array of papers.paperid values.
  $query = "select distinct papers.paperid from $tablestr where $condstr";

  $result = get_list_query($query);

  // display the results of the query
  if (is_array($result)) {
    $cnt = count($result);
    $i = 0;
    $sr = $fr + 1;
    $er = ($cnt < ($fr + 60)) ? $cnt : $fr + 60;
    if ($cnt == 1) {
      echo "<p>Found one matching paper, displaying ";
    }
    else {
      echo "<p>Found $cnt matching papers, displaying $sr to $er ";
    }
    if ($cnt < 20) {
      echo "verbosely";
    }
    echo ":</p>";
    foreach ($result as $elem) {
      if ($i >= $fr && $i < $er) {
        $paperid = $elem['paperid'];
        $paper = get_paper($paperid);
        if ($cnt < 20) // only display verbosely if there are few records matching...
        {
          display_paper_verbose($paper);
        }
        else {
          display_paper_link(0, $paper);
        }
      }
      if ($i >= $er) {
        break;
      }
      $i++;
    }
  }
  else {
    do_para("Sorry, No results found for query.");
  }
}

{
  echo "<br>";

  $fr = isset($_POST['fr']) ? $_POST['fr'] : 0;
  $num = isset($_POST['num']) ? $_POST['num'] : 1;
  if (isset($_POST['moreButton']) && $_POST['moreButton'] == "More") {
    $num = $num + 1;
  }
  if (isset($_POST['fewButton']) && $_POST['fewButton'] == "Fewer") {
    $num = $num - 1;
  }

  echo "<form method=\"post\" action=\"search.php\">\n";
  echo "<input type=\"hidden\" name=\"num\" value=\"$num\">\n";
  echo "<input type=\"hidden\" name=\"fr\" value=\"$fr\">\n"; // first record
  echo "<table>";
  for ($i = 1; $i <= $num; $i++) {
    $v = isset($_POST['text'][$i]) ? $_POST['text'][$i] : "";
    $w = isset($_POST['what'][$i]) ? $_POST['what'][$i] : "";

    echo "<tr>";
    echo "<td>\n";
    echo "<select name=\"what[$i]\">\n";
    echo "<option value=\"1\"";
    if ($w == 1) {
      echo " selected";
    }
    echo ">Title</option>\n";
    echo "<option value=\"2\"";
    if ($w == 2) {
      echo " selected";
    }
    echo ">Author Surname</option>\n";
    echo "<option value=\"3\"";
    if ($w == 3) {
      echo " selected";
    }
    echo ">Author Firstname</option>\n";
    echo "<option value=\"4\"";
    if ($w == 4) {
      echo " selected";
    }
    echo ">Abstract</option>\n";
    echo "<option value=\"5\"";
    if ($w == 5) {
      echo " selected";
    }
    echo ">Ref No</option>\n";
    echo "<option value=\"6\"";
    if ($w == 6) {
      echo " selected";
    }
    echo ">Pub. Date</option>\n";
    echo "<option value=\"7\"";
    if ($w == 7) {
      echo " selected";
    }
    echo ">ISBN</option>\n";
    echo "<option value=\"8\"";
    if ($w == 8) {
      echo " selected";
    }
    echo ">Has File Attachments</option>\n";
    echo "</select>";
    echo "</td><td><input type=\"text\" name=\"text[$i]\" size=\"60\" maxlength=\"120\" value=" . sqlvalue($v) . "></td>\n";
  }
  echo "<tr><td valign=\"bottom\" rowspan=\"2\" align=\"left\">";
  echo "<input type=\"submit\" name=\"goButton\" value=\"Search\"></td></tr>\n";
  echo "<tr><td valign=\"bottom\"  align=\"right\"><input type=\"submit\" name=\"moreButton\" value=\"More\">\n";
  if ($num > 1) {
    echo "<input type=\"submit\" name=\"fewButton\" value=\"Fewer\">";
  }
  echo "</td></tr>\n";
  echo "</table></form>";

}

do_html_footer();